tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/SymbStateDecoder.scala

```diff
@@ -156,6 +132,49 @@ class SymbStateDecoder(solverContext: SolverContext, rewriter: SymbStateRewriter
       throw new RewriterException("Don't know how to decode the cell %s of type %s".format(cell, cell.cellType), NullEx)
   }
 
+  private def decodeOldRecordToTlaEx(arena: Arena, cell: ArenaCell, recordT: RecT1): TlaEx = {
+    def exToStr(ex: TlaEx): TlaStr = ex match {
+      case ValEx(s @ TlaStr(_)) => s
+      case _                    => throw new RewriterException("Expected a string, found: " + ex, ex)
+    }
+
+    // Note that the domain may have fewer fields than the record type is saying.
+    // This comes from the fact that we can extend a record with a richer type.
+    val domCell = arena.getDom(cell)
+    val dom = decodeSet(arena, domCell).map(exToStr)
+    val fieldValues = arena.getHas(cell)
+    val keyList = recordT.fieldTypes.keySet.toList
+
+    def eachField(es: List[TlaEx], key: String): List[TlaEx] = {
+      if (!dom.contains(TlaStr(key))) {
+        es // skip
+      } else {
+        val index = keyList.indexOf(key)
+        val valueCell = fieldValues(index)
+        tla.str(key).typed() +: decodeCellToTlaEx(arena, valueCell) +: es
+      }
+    }
+
+    val keysAndValues = keyList.reverse.foldLeft(List[TlaEx]())(eachField)
+    if (keysAndValues.nonEmpty) {
+      OperEx(TlaFunOper.enum, keysAndValues: _*)(Typed(recordT))
+    } else {
+      logger.error(
+          s"Decoder: Found an empty record $cell when decoding a counterexample, domain = $domCell. This is a bug.")
+      // for debugging purposes, just return a string
+      tla.str(s"Empty record domain $domCell").typed()
```

Comment:
Why shouldn't this raise an exception or return a value indicating the error explicitly?

Reply:
Yeah. It probably should. This is still `decodeOldRecordToTlaEx`. We will remove it very soon!